File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/ledger/ManageLedger.java
package com.decrux.pocketr.api.services.ledger;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.dtos.BalanceDto;
import com.decrux.pocketr.api.entities.dtos.CreateTransactionDto;
import com.decrux.pocketr.api.entities.dtos.PagedTransactionsDto;
import com.decrux.pocketr.api.entities.dtos.TransactionDto;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface ManageLedger {

    TransactionDto createTransaction(CreateTransactionDto dto, User creator);

    PagedTransactionsDto listTransactions(
        User user,
        String mode,
        UUID householdId,
        LocalDate dateFrom,
        LocalDate dateTo,
        UUID accountId,
        UUID categoryId,
        int page,
        int size
    );

    List<BalanceDto> getAccountBalances(List<UUID> accountIds, LocalDate asOf, User user, UUID householdId);

    BalanceDto getAccountBalance(UUID accountId, LocalDate asOf, User user, UUID householdId);
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/services/ledger/ManageLedger.kt
package com.decrux.pocketr_api.services.ledger

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.dtos.BalanceDto
import com.decrux.pocketr_api.entities.dtos.CreateTransactionDto
import com.decrux.pocketr_api.entities.dtos.PagedTransactionsDto
import com.decrux.pocketr_api.entities.dtos.TransactionDto
import java.time.LocalDate
import java.util.UUID

interface ManageLedger {
    fun createTransaction(
        dto: CreateTransactionDto,
        creator: User,
    ): TransactionDto

    fun listTransactions(
        user: User,
        mode: String?,
        householdId: UUID?,
        dateFrom: LocalDate?,
        dateTo: LocalDate?,
        accountId: UUID?,
        categoryId: UUID?,
        page: Int,
        size: Int,
    ): PagedTransactionsDto

    fun getAccountBalances(
        accountIds: List<UUID>,
        asOf: LocalDate,
        user: User,
        householdId: UUID?,
    ): List<BalanceDto>

    fun getAccountBalance(
        accountId: UUID,
        asOf: LocalDate,
        user: User,
        householdId: UUID?,
    ): BalanceDto
}
